src/helpers/recompute_zones.ts

```diff
@@ -0,0 +1,398 @@
+    i--
+  ) {
+    if (
+      deepEqualsArray(
+        profiles.get(profilesStartingPosition[i])!,
+        profiles.get(profilesStartingPosition[i - 1])!
+      )
+    ) {
+      profiles.delete(profilesStartingPosition[i]);
+      profilesStartingPosition.splice(i, 1);
+    }
+  }
+}
+
+function constructZonesFromProfiles(
+  profilesStartingPosition: number[],
+  profiles: Map<number, number[]>
+): UnboundedZone[] {
+  const mergedZone: UnboundedZone[] = [];
+  let stackZones: UnboundedZone[] = [];
+  for (let colIndex = 0; colIndex < profilesStartingPosition.length; colIndex++) {
+    const left = profilesStartingPosition[colIndex];
+    const profile = profiles.get(left)!;
+    if (!profile || profile.length === 0) {
+      mergedZone.push(...stackZones);
+      stackZones = [];
+      continue;
+    }
+
+    let right = profilesStartingPosition[colIndex + 1];
+    if (right !== undefined) {
+      right--;
+    }
+
+    const colZones: UnboundedZone[] = [];
+    for (let i = 0; i < profile.length; i += 2) {
+      const top = profile[i];
+      let bottom = profile[i + 1];
+      if (bottom !== undefined) {
+        bottom--;
+      }
+      const colZone: UnboundedZone = {
+        top,
+        left,
+        bottom,
+        right,
+        hasHeader: (bottom === undefined && top !== 0) || (right === undefined && left !== 0),
+      };
+
+      let findCorrespondingZone = false;
+      for (let j = stackZones.length - 1; j >= 0; j--) {
+        const stackZone = stackZones[j];
+        if (stackZone.top === colZone.top && stackZone.bottom === colZone.bottom) {
+          stackZone.right = colZone.right;
+          stackZones.splice(j, 1);
+          colZones.push(stackZone);
+          findCorrespondingZone = true;
+          continue;
+        }
+      }
+      if (!findCorrespondingZone) {
+        colZones.push(colZone);
+      }
+    }
+
+    mergedZone.push(...stackZones);
+    stackZones = colZones;
+  }
+  mergedZone.push(...stackZones);
+  return mergedZone;
+}
+
+function binaryPredecessorSearch(arr: number[], val: number, start = 0, matchEqual = true) {
+  let end = arr.length - 1;
+  let result = -1;
+
+  while (start <= end) {
+    const mid = Math.floor((start + end) / 2);
+    if (arr[mid] === val && matchEqual) {
+      return mid;
+    } else if (arr[mid] < val) {
+      result = mid;
+      start = mid + 1;
+    } else {
+      end = mid - 1;
+    }
+  }
+
+  return result;
+}
+
+function binarySuccessorSearch(arr: number[], val: number, start = 0, matchEqual = true) {
+  let end = arr.length - 1;
+  let result = arr.length;
+  while (start <= end) {
+    const mid = Math.floor((start + end) / 2);
+    if (arr[mid] === val && matchEqual) {
+      return mid;
+    } else if (arr[mid] > val) {
+      result = mid;
+      end = mid - 1;
+    } else {
+      start = mid + 1;
+    }
+  }
+  return result;
+}
+
+function deepEqualsArray(arr1: number[], arr2: number[]): boolean {
```

Comment:
Can't you use deepEquals ? I don't think this is really faster

Reply:
here we know that we are working with arrays,
we have the discriminating condition: 
```js
  if (arr1.length !== arr2.length) {
    return false;
  }
``` which allows many cases to be quickly eliminated.
Isn't that faster??